File: androidApp/src/main/java/com/ziina/multiplatformpath/android/MultiPlatformView.kt
package com.ziina.multiplatformpath.android

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.unit.dp
import com.ziina.multiplatformpath.MultiPlatformDrawer
import com.ziina.multiplatformpath.PlatformDrawer
import com.ziina.multiplatformpath.models.DrawingRect
import com.ziina.multiplatformpath.models.LineTo
import com.ziina.multiplatformpath.models.MoveTo

@Composable
fun MultiPlatformView() = Canvas(
    modifier = Modifier
        .size(200.dp)
        .background(Color.Black),
    onDraw = {
        val path = Path()

        PlatformDrawer.getDrawingFunctions(DrawingRect(size.width, size.height))
            .forEach { drawingFunction ->
                when (drawingFunction) {
                    is LineTo -> path.lineTo(drawingFunction.x, drawingFunction.y)
                    is MoveTo -> path.moveTo(drawingFunction.x, drawingFunction.y)
                }
            }

        drawPath(path, Color.Red)
        path.close()
    }
)


// This also works fine
@Composable
fun MultiPlatformViewMultiPlatformDraw() = Canvas(
    modifier = Modifier
        .size(200.dp)
        .background(Color.White),
    onDraw = {
        val path = Path()
        MultiPlatformDrawer.drawOnPath(path, DrawingRect(size.width, size.height))
        drawPath(path, Color.Blue)
        path.close()
    }
)

File: iosApp/iosApp/MultiPlatformView.swift
import SwiftUI
import shared

struct MultiPlatformView: Shape {
    let platformDrawer = PlatformDrawer()
    
    func path(in rect: CGRect) -> Path {
        var path = Path()

        platformDrawer
            .getDrawingFunctions(drawingRect: DrawingRect(width: Float(rect.width), height: Float(rect.height)))
            .forEach { drawingFunction in
                switch(drawingFunction){
                case let move as MoveTo: path.move(to: CGPoint(x: Double(move.x), y: Double(move.y)))
                case let line as LineTo: path.addLine(to: CGPoint(x: Double(line.x), y: Double(line.y)))
                default: break
                }
            }

        return path
    }

}

struct MultiPlatformViewMultiPlatformDraw: Shape {
    let multiplatformDrawer = MultiPlatformDrawer()
    
    // 😭 doesn't work. Draws nothing
    func path(in rect: CGRect) -> Path {
        var cgPath = CGMutablePath()
        multiplatformDrawer.drawOnPath(path: &cgPath, drawingRect:  DrawingRect(width: Float(rect.width), height: Float(rect.height)))
        return Path(cgPath)
    }

}

